### DISCUSSION TITLE
provide function to turn a ResultSet in a List<Columns>

### DISCUSSION THREAD

**Discussion by robstoll** [robstoll/atrium#1344] [Ideas]
As pointed out by @vlsi 
> Well, I have two issues [with using Atrium in conjunction with ResultSet]:
> 1) `ResultSet` column accessor has two axes: column name (or position), and accessor method (getInt, getString, or even Kotlin extension method for retrieving a custom type). I believe that makes impractical trying to come up with a fancy extractor like `expect(rs).column("name")`. Those extractors will be non-trivial to use, and it won't make the output dramatically better
> 2) It would be nice to have an API to peek on the current state of `Expect<T>` to skip subsequent soft assertions when they make no sense. For instance, the first thing I check is `rs.next()`, and I need skip processing when there's no data in the resultset. What is the idiomatic way to do that? `expect(rs.next()).toEqual(true)`? How do I know if the expectation holds?
> For now, I use `if (!rs.next()) { _logic.append(.."no rows found for.."); continue}`, however, it looks weird

As long as one does not intend to get data via `get...Stream` functions, then it could be OK to load all data eagerly and turn the ResultSet into a `List<Columns>` where Atrium performs the mapping and `Columns` is a data class provided by Atrium.
I would then provide an extension function `get<T>(string|Int)`.
The benefit would be that one can re-use all the functionality we provide for Iterable, i.e. all toContain functions as well as `toHaveElementsAndAll` etc.

An example:
```kotlin
val rs = //select productName, price
expect(rs).asList().toHaveElementsAndAll {
  get<String?>("productName").notToBeNullOrBlank()
  get<Int>("price").toBeGreaterThan(5)
}
```
And the output then could look as follows:
```
I expected subject: ResultSet for table Xy with 10 rows and 2 columns
◆ elements need all: 
    » column productName: 
      » to be an instance of type: String? 
      » not to be null or blank
    » column price:
       » to be an instance of type: Int 
       » to be greater than: 5        
    ❗❗ following elements were mismatched: 
       ⚬ index 4: Columns(productName: "foo", price: 1)
```
If the naming in the reporting bothers us too much, then we cold improve it later on IMO, i.e. `elements need all` could be turned into `rows need all` and `index 4` would be `row 4`. 

The benefit of using this approach is that a developer sees more context (i.e. all columns not just the one where a comparison fails) giving good hints why something fails.

Another example:
```
val rs = // select id, min, max
expect(rs).asList().toContainOnlyInAnyOrder(
  { 
    get<UUID>("id").toEqual(expectedId1)
    get<BigDecimal>("min").toBeGreaterThan(expectedMinimalMin)
  },
  {
    get<UUID>("id").toEqual(expectedId2)
    get<BigDecimal>("max").toBeLessThan(expectedMaximalMax)
  }
)
```
Would look like something along the line of in reporting:
```
I expected subject: ResultSet for table Xy with 3 rows and 3 columns
◆ to contain only, in any order: 
  ✔ an element which needs:
      » column id: 
         » to be an instance of type: UUID
         » to equal: f3395854-2b3c-44ad-8fa9-d8fa8de22df2
      » column min
        » to be an instance of type: BigDecimal
        » to be greater than: BigDecimal(2.3)
  ✔ an element which needs:
      » column id: 
         » to be an instance of type: UUID
         » to equal: f3395854-2b3c-44ad-8fa9-d8fa8de22df3
      » column max
        » to be an instance of type: BigDecimal
        » to be less than: BigDecimal(45.78)
  ❗❗ additional elements detected
     ⚬ Columns(id: f3395854-2b3c-44ad-8fa9-d8fa8de22df1, min: BigDecimal("1.4343"), max: BigDecimal("51.2")
```

@vlsi WDYT would that be valuable for you and would it solve your issue regarding `peek on the current state`?

**Comment by vlsi** [comment]
The above sounds reasonable, however, it would probably require some data-mapping and conversion when accessing the DB.
For instance, if the database returns `max` as `varchar`, you would probably store it in `Columns` as `String`, and if the expectation is `get<BigDecimal>(max)`, then you'll need to convert the thing somehow.
It would be extremely hairy for timestamp-like types.

Then, queries often require parameters.
So the typical use is

```kotlin
val db = ...
db.query(
    """
    select ... from ... where id = ? and name = ?
    """.trimIndent(),
    id,
    name,
) { rs ->
  ...
}
```

That use pattern works well for encapsulating parameter conversion logic (e.g. the way different datatypes should be converted to DB), and it ensures `query` method closes resultset automatically, so it does not leak.

If you start asserting on `ResultSet`, then you somehow force the user to "leak" resultset and deliver it to Atrium. That sounds like bad practice.

Here's another sample:

```kotlin
expectSoftly {
    data class MappingApplyResult(
        val explainPlanUid: Id<ExplainPlan, Uid>?, //
        val throwable: Throwable? // null on success
    )
    // explainMappingApply returns Map<String, MappingApplyResult>
    // where the key is database name and the value is the result at that database
    expect(configManager.explainMappingApply(failOnError = false)) {
        notToBeEmpty()
        // There should probably be Expect<Map<K, V>>.toHaveElementsAndAll(...Expect<Map.Entry<K, V>>...
        asEntries {
            // I can assert key and value independently, however, what I need is
            // to load "value" for every "key"
            toHaveElementsAndAll {
                value.throwable.notToEqualNull()
                    .message
                    .because("test configured the system to fail") {
                        toContain("break explain plan to check if it properly shows FAILED status")
                    }

                // How would I load the data?
                // I need to access both key and value, so the only way out is "feature(...)" :-/
                // However, it results in a quite involved "feature"
                feature({
                    val dbName = key
                    val applyPlan = value
                    // assertManager is an injected service for database asserts and queries
                    assertManager.at(dbName) { qe ->
                        qe.query(
                            """
                            select state
                              from session_traces t
                             where t.explain_plan_uid = ?
                            """.trimIndent(),
                            { rs ->
                                // I don't quite like the idea of manually serializing resultSet data
                                // for the sole purpose of expecting it later
                                while (rs.next()) {
                                    expect(rs.getString("state")).toEqual("FAILED")
                                }
                            },
                            applyPlan.explainPlanUid.id
                        )
                    }
                }) {
                    // empty? all expectations are in "feature" above
                }
            }
        }
```

This time the query is trivial, and in theory, I can return "state" as a feature result, and then run an expectation.
However, it sounds scary if "one should map all the rows to objects in order to execute assertions".

I wonder what you think of having something like 

```kotlin
    fun <T> Expect<T>.unwrapSubject(
        description: String,
        action: T.() -> Unit
    ): Expect<T> =
        apply {
            when (val subject = _logic.maybeSubject) {
                is Some -> subject.value.action()
                is None -> _logic.createAndAppend(Untranslatable(description), null) { false }
            }
        }
```

so the code could be like

```kotlin
toHaveElementsAndAll {
    value.throwable.notToEqualNull()
        .message
        .because("test configured the system to fail") {
            toContain("break explain plan to check if it properly shows FAILED status")
        }

    unwrapSubject("Session in database should be in FAILED state") {
        val dbName = key
        val applyPlan = value
        // assertManager is an injected service for database asserts and queries
        assertManager.at(dbName) { qe ->
            qe.query(
                """
                select state
                  from session_traces t
                 where t.explain_plan_uid = ?
                """.trimIndent(),
                { rs ->
                    // It looks like I can use expect(...) for expecting the columns
                    // TODO: add an expectation that the query returns exactly one row
                    while (rs.next()) {
                        expect(rs.getString("state")).toEqual("FAILED")
                    }
                },
                applyPlan.explainPlanUid.id
            )
        }
    }
}
```

Even though `"Session ${value.explainPlanUid} in database $key should be in FAILED state"` duplicates the logic below, it seems to allow reusing existing APIs without converting results for every assertion.

**Comment by robstoll** [comment]
Just that you know, I have never worked with ResultSet so far and I just made some assumptions without really looking at the documentation.

> For instance, if the database returns max as varchar, you would probably store it in Columns as String, and if the expectation is get<BigDecimal>(max), then you'll need to convert the thing somehow.

I assumed `rs.getBigDecimal` would throw an SqlException in case the column is a (N)(Var)Char. If my assumption is correct, then I would let the expecation fail in such a case, mentioning that the column had the wrong type. But it sounds like it somehow does a conversion, correct? 

> If you start asserting on ResultSet, then you somehow force the user to "leak" resultset and deliver it to Atrium. That sounds like bad practice.

I guess I see your point, but I am not entirely sure. Because if you would use the mapping in your example then you would not leak more than now. But I assume you refer to the case if you would have passed rs somehow as subject, right? 
I agree that this is smelly and we would need to implement it properly so that nothing is leaked.


> // There should probably be Expect<Map<K, V>>.toHaveElementsAndAll(...Expect<Map.Entry<K, V>>...

So far I have not seen many cases where we need this functionality on Map. That's why there is currently no shortcut function and one has to go over `asEntries`

> However, it sounds scary if "one should map all the rows to objects in order to execute assertions".

That was my "fear", that if someone works with ResultSet the eager loading all rows and map into a data class approach might not be what such a person is used to, because in the business code it usually matters performance wise. I guess it does not matter in tests unless you want a fail-fast behaviour, i.e. stop at the first row which does not hold all expectations. For such a case my suggested approach is not feasible.
However, in all examples you have shown so far, you seem to want to go over all rows (not fail fast) and report all rows which do not hold the expectations. I guess the mapping is still very small compared to IO involved in loading the data in such cases. It might be noticeable if we load a lot of data and do it in many tests (which are not parallelised). Say a test needs 20s instead of 18s due to the additional mapping. If you have 3600 such tests, then it would take 20h instead of 18h. 2h might seem a lot but I would argue, your real problem here is the lack of parallelisation and not the 2s mapping overhead per test. 
The question remaining is though, if you want to do such a paradigm shift in tests or if you prefer to have the same approach you use in your business logic in tests as well?

> I wonder what you think of having something like

I am not yet sure if we should put it on the API level as it will pop up for any `Expect<T>`. Maybe we only add it on FeatureExpect. 
Basically I would like to avoid that a user sees it when he uses `expect(person).unwrapSubject` because it is clear that at this point the subject always exists.
We can definitely add it on the logic level as first step, not sure yet about the naming though. Would you mind to open another discussion for this specific point?

>  // TODO: add an expectation that the query returns exactly one row

This would be covered with the asList approach:
```kotlin
qe.query(
    """
    select state
      from session_traces t
     where t.explain_plan_uid = ?
    """.trimIndent(),
    { rs ->
        expect(rs).asList().toContainExactly {
            get<String>("state").toEqual("FAILED")
        }
    },
    applyPlan.explainPlanUid.id
)
```

A flash thought, maybe it would make more sense to type the columns during the `asList` call, this way we would not need to repeat it, we could use functions which ResultSet provides and the reporting could look a bit nicer as well.
```
expect(rs).asList({ getObject<QueryState>("state") }).toContainExactly{
   column1.toEqual(QueryState.FAILED) // column1 is typed as QueryState
}
```
Yet, if you deal with many columns, then using column1...n is not that readable.

<hr/>

Last but not least, in the end I (we) won't add any ResultSet functionality which you don't intend to use. You are the only user I am aware of using Atrium in combination with ResultSet so far, so we should focus on functionality which will actually improve your work with Atrium :slightly_smiling_face: 
I was hoping re-using the expectation function of Iterable would already help you out but if we need more tailored expectation functions for Resultset then we should not stop at the current functionality.

**Reply by vlsi** [reply]
>I assumed rs.getBigDecimal would throw an SqlException in case the column is a (N)(Var)Char. If my assumption is correct, then I would let the expecation fail in such a case, mentioning that the column had the wrong type. But it sounds like it somehow does a conversion, correct?

There is a bunch of standard conversions, so implementing a caching `ResultSet` is not trivial.

> The question remaining is though, if you want to do such a paradigm shift in tests or if you prefer to have the same approach you use in your business logic in tests as well?

Fetching all results in memory is not an issue for me.
The issue is that I was not prepared for temporarily wrapping the results in Kotlin objects for the sole reason of writing expectations later.

>```
>expect(rs).asList({ getObject<QueryState>("state") }).toContainExactly{
>   column1.toEqual(QueryState.FAILED) // column1 is typed as QueryState
>}
>```
>Yet, if you deal with many columns, then using column1...n is not that readable.

That exactly summarizes why I am hesitant to buffering results.
If we buffer results into a generic `List<List<Cell>>`-like structure, then expectations would become like `get(42).toEqual(...)`.
If I create a properly named class, then the expectations would become better (they would be able to use proper field names), however, I am not sure what would be code reuse there.

>The question remaining is though, if you want to do such a paradigm shift in tests or if you prefer to have the same approach you use in your business logic in tests as well?

99% of my "business logic" is written in PL/SQL. I use Kotlin+TestNG+Guice+Atrium for component/integration tests.
The test I shared so far were for system state verification (~control plane), not for business data verification (~data plane).

In the ideal world, I would probably have Kotlin wrapper objects for all the entities (including database-state ones), and I should have written tests like

```kotlin
expect(configManager.explainMappingApply(failOnError = false)) {
    loadPlanState(db=key, value=planId) { // <-- access key and value from the Map somehow, and use database connection from a context receiver or whatever
        state // <-- KSP-generated accessor
            .because("the plan was doomed to fail because there was a targeted misconfiguration, so its state should be FAILED")
            .toEqual(State.FAILED)
    }
}
```

However, I'm not there yet.
There's no KSP plugin that would generate accessors like the following for every class I create, and asking the team to generate them manually, well, sucks. Especially, if you consider the team is mostly Oracle PL/SQL developers.

```
data class MappingApplyResult(
    val explainPlanUid: Id<RdblExplainPlan, Uid>,
    val throwable: Throwable?
)

// Below accessors could have generated automatically
val Expect<MappingApplyResult>.throwable: FeatureExpect<MappingApplyResult, Throwable?>
    get() = feature({ throwable })

val Expect<MappingApplyResult>.explainPlanUid: FeatureExpect<MappingApplyResult, Id<ExplainPlan, Uid>>
    get() = feature({ explainPlanUid })
```

----

> Basically I would like to avoid that a user sees it when he uses `expect(person).unwrapSubject` because it is clear that at this point the subject always exists.

There are cases when Atrium executes assertion creator on a `None` subject with "no apparent reason" 😉 : https://github.com/robstoll/atrium/blob/65227add3c31604a64cf4ef86671eadc78be3bde/logic/atrium-logic/src/commonMain/kotlin/ch/tutteli/atrium/logic/creating/collectors/AssertionsOptionExplantoryExtensions.kt#L34-L36

---

> Because if you would use the mapping in your example then you would not leak more than now. But I assume you refer to the case if you would have passed rs somehow as subject, right?

I would probably try adding `Expect<ResultSet>.toHaveElementsAndAll(assertionCreatorOrNull: (Expect<...>.() -> Unit)?): Expect<ResultSet> =` to my project and see how it goes, however, I did not get there because, well, it turns out I can't easily access the subject, and there are cases when the subject is unexpected `None`.

I see how `None` subject allows Atrium to print all the messages, including non-executed ones. It is really powerful, however, it is sad it is not compatible with gradual adoption where one starts with using minimal subset of Atrium, and eventually adds targeted feature extractors.

**Reply by robstoll** [reply]
> There are cases when Atrium executes assertion creator on a None subject with "no apparent reason" wink : 

This shouldn't be an issue as the subject can also be None if a transformation or feature extraction did not work out as expected.
For instance:
```
val i: Any = "1"
expect(i).toBeAnInstanceOf<Int> {
  // maybeSubject is now None
  unwrapSubject("....") {
  }
}
```
The message you supply to unwrapSubject should not be appended via `_logic.createAndAppend(Untranslatable(description), null) { false }` but an explanatory Assertion should be used instead `assertionBuilder.explanatory.withExplanation(description).build()`
I guess this will solve your issue.

> In the ideal world, I would probably have Kotlin wrapper objects for all the entities (including database-state ones), and I should have written tests like

So let's target this goal. 
1. Generating the data classes based on your tables could be done via some entity framework or the like -- I can recommend Jooq as it allows to generate kotlin data classes and it would also allow to do the mapping => nothing to do in Atrium
2. > access key and value from the Map somehow 
   
    => we need to introduce something like unwrapSubject
3. > state // <-- KSP-generated accessor

    you already created a corresponding issue

5. > .because("the plan was doomed to fail because there was a targeted misconfiguration, so its state should be FAILED")
    > .toEqual(State.FAILED)
    
   => I suggest we add a because parameter to each expectation function, i.e. you would write it as follows:
       `toEqual(State.FAILED, because = "the plan was doomed to fail because there was a targeted misconfiguration, so its state should be FAILED")` WDYT?

**Reply by vlsi** [reply]
> I can recommend Jooq as it allows to generate kotlin data classes

Jooq requires a license for Oracle DB, and it is tough in my case.
Of course "generating data classes" is out of scope for Atrium, so I would try going with a solution that does not require bringing entity framework.

`unwrapSubject` looks like a reasonable trade-off, however, I need to play with it more.

>I suggest we add a because parameter to each expectation function, i.e. you would write it as follows:
toEqual(State.FAILED, because = "the plan was doomed to fail because there was a targeted misconfiguration, so its state should be FAILED") WDYT?

It would be nice to have that. It sounds scary to manually add a parameter to all the functions though. Especially, if we ever consider to rollback that change or adding one more parameter.
It might be we could autogenerate expectation functions with `because: String` parameter. The drawback of "generating extra to... functions" would be they would clutter autocomplete menu.

**Reply by robstoll** [reply]
We don't have that many expectation function yet, so I think we are good. What I am more concerned about is if we will run into ambiguity issues in the future similar to the one you had with the CallerArguments compiler plugin
